### DISCUSSION TITLE
Syncing pwd with Fern root dir

### DISCUSSION THREAD

**Discussion by iamvoidcoder** [lambdalisue/vim-fern#322] [Q&A]
Hi

Is there an easy way to somehow auto-sync pwd with Fern's root dir? Via a custom Fern autocmd event perhaps? So that if you drill in/out in the Fern's file system tree whenever the root dir changes to automatically perform :cd <current_fern_root>?

The reason is quite simple. It is sort of hard to manually keep these two in sync. It would be much easier if e.g. after drilling in into a folder in Fern (Enter) or moving up a level (Backspace) it also changed pwd so that any relative path commands or plugins that rely on pwd (such as e.g. FZF) would work relative to the current Fern root dir.

If there is a way to somehow hook a custom script handler for the Fern root change events and obtain the current root path I can see this sorting both the pwd syncing and my other question about preserving drawer state.

Thanks

**Comment by lambdalisue** [comment]
Maybe https://github.com/lambdalisue/fern.vim/wiki/Tips#perform-tcd-on-the-root-node-after-enter-or-leave-action would help you.

**Reply by iamvoidcoder** [reply]
Thanks. I just tried that and it doesn't seems to do anything. I don't think the action enter and leave get invoked unless I'm doing something wrong. I just plugged this in into my nvim init script and I can't see any feedback when I leave or enter dirs:

```
function! s:init_fern() abort

  nmap <buffer> <Plug>(fern-my-enter-and-tcd)
        \ <Plug>(fern-action-enter)
        \ <Plug>(fern-wait)
        \ echo "Going down!"

  nmap <buffer> <Plug>(fern-my-leave-and-tcd)
        \ <Plug>(fern-action-leave)
        \ <Plug>(fern-wait)
        \ echo "Going up!"

endfunction

augroup fern-custom
  autocmd! *
  autocmd FileType fern call s:init_fern()
augroup END

```

**Reply by lambdalisue** [reply]
You don't see any feedback because the code is completely wrong. Read `:help map`

**Reply by lambdalisue** [reply]
Additionally, I may mis-understood your want. Please explain the behavior you want with simple step by step procedure (while it's a bit tough work to understand/write English for me)

**Reply by iamvoidcoder** [reply]
All I want is just my own script function to be called whenever Fern root directory changed and somehow obtain the new root path in that function. From there I should be able to achieve everything I need on my own. 

Bear in mind that not everyone is a vim script ninja (and thankfully doesn't have to be, let's leave this to the developers!). I certainly used some basic key mapping before but this `nmap <buffer> <Plug>` stuff is beyond my understanding.

Can you please show how to hook in such a function and simply print the new root path every time it changes? Something like:

function! onFernRootChanged(new_root_path)
      echo "Fern root changed to: " .. new_root_path
endfunction

Thanks

**Reply by lambdalisue** [reply]
Add the followings to your vimrc

```vim
function! s:hello() abort
  echomsg "HELLO"
endfunction

nnoremap <Plug>(my-hello) :<C-u>call <SID>hello()<CR>
```

Then use the `<Plug>(my-hello)` to check. Note that you need to execute `:messages` to confirm after you hit the key.

```vim
function! s:init_fern() abort
  nmap <buffer> <Plug>(fern-my-enter-and-tcd)
        \ <Plug>(fern-action-enter)
        \ <Plug>(fern-wait)
        \ <Plug>(my-hello)
  nmap <buffer> <Plug>(fern-my-leave-and-tcd)
        \ <Plug>(fern-action-leave)
        \ <Plug>(fern-wait)
        \ <Plug>(my-hello)
  " ...
endfunction

augroup fern-custom
  autocmd! *
  autocmd FileType fern call s:init_fern()
augroup END
```

Well, I noticed that your code does not include mappings to `<Plug>(fern-my-enter-and-tcd)` nor `<Plug>(fern-my-leave-and-tcd)`. If you haven't mapped those, add the followings in `s:init_fern()`

```vim
  nmap <buffer><nowait> <CR> <Plug>(fern-my-enter-and-tcd)
  nmap <buffer><nowait> <BS> <Plug>(fern-my-leave-and-tcd)
```

**Reply by iamvoidcoder** [reply]
Interesting. I copy pasted your example code to vimrc as is. I can't really see it printing HELLO when I enter or leave a directory in the Fern drawer. Even with the two `<CR> and <BS>` mapping lines. Have you actually tried that example? Something is still missing :( 

Also how do I access / print the current Fern root path in that hello() function? Without knowing what Fern root path changed to this exercise is kinda useless.

BTW I have just noticed every time I enter `<CR>` or leave `<BS>` a directory a new instance of Fern buffer gets created! I can see them all with :ls! I mean, after browsing around for a while I end up with a ton of unneeded background buffers. Is this really expected behaviour and is there a way to keep a single buffer?

Thanks

**Reply by lambdalisue** [reply]
>Interesting. I copy pasted your example code to vimrc as is. I can't really see it printing HELLO when I enter or leave a directory in the Fern drawer. Even with the two <CR> and <BS> mapping lines. Have you actually tried that example? Something is still missing :(

I think there were enough information. I've tried with the following minimal vimrc created from the previous post by executing `nvim -u ~/vimrc.min` and `HELLO` is printed when I hit `<CR>` or `<BS>` on fern buffer. Note that I needed to hit `:messages` to see that.


```vim
if exists('+compatible') && &compatible
  set nocompatible
endif

set packpath=
set runtimepath^=~/ghq/github.com/lambdalisue/fern.vim

filetype plugin indent on
syntax on

function! s:hello() abort
  echomsg "HELLO"
endfunction

nnoremap <Plug>(my-hello) :<C-u>call <SID>hello()<CR>
nmap H <Plug>(my-hello)

function! s:init_fern() abort
  nmap <buffer> <Plug>(fern-my-enter-and-tcd)
        \ <Plug>(fern-action-enter)
        \ <Plug>(fern-wait)
        \ <Plug>(my-hello)
  nmap <buffer> <Plug>(fern-my-leave-and-tcd)
        \ <Plug>(fern-action-leave)
        \ <Plug>(fern-wait)
        \ <Plug>(my-hello)
  nmap <buffer><nowait> <CR> <Plug>(fern-my-enter-and-tcd)
  nmap <buffer><nowait> <BS> <Plug>(fern-my-leave-and-tcd)
endfunction

augroup fern-custom
  autocmd! *
  autocmd FileType fern call s:init_fern()
augroup END
```

> Also how do I access / print the current Fern root path in that hello() function? Without knowing what Fern root path changed to this exercise is kinda useless.

Yes it's useless. I made that function because you've used `echo "Going up"` in previous post.

You have to replace `<Plug>(my-hello)` to mappings described on https://github.com/lambdalisue/fern.vim/wiki/Tips#perform-tcd-on-the-root-node-after-enter-or-leave-action

> BTW I have just noticed every time I enter <CR> or leave <BS> a directory a new instance of Fern buffer gets created! I can see them all with :ls! I mean, after browsing around for a while I end up with a ton of unneeded background buffers. Is this really expected behaviour and is there a way to keep a single buffer?

No. Fern is buffer oriented plugin thus fern create fresh buffers for each root (actually, for each call). `ls!` list unlisted buffer as well. Use `ls` instead of `ls!` for daily use.

**Reply by iamvoidcoder** [reply]
Please don't get upset. Just remember not everyone is a vimscript master. Yet we still want to use your plugin nonetheless!

When I pasted your last example it worked OK, I'm now getting HELLO printed. I'm almost there. Now I only need to find how to access current Fern root path (for example print current root path instead of HELLO).

About ls!

ls doesn't print plugin buffers. I only started looking into unlisted buffers because I noticed that after I spend some time navigating file system in Fern via enter (`<CR>`) & leave (`<BS>`) the amount of RAM used by nvim is starting to grow up a lot. 

What I discovered is that every time I enter into a folder or leave a folder this appears to be creating a new fern buffer instance and makes it active. Previous Fern buffers (all previous root folders I visited) remain open and I can switch to them via `:b<number>`. This sounds more like a memory leak to me.

There are more than 40 unlisted  `fern://drawer:1/file:///...`  buffers in my current session.

**Reply by lambdalisue** [reply]
Well, I didn't... I don't know how to say Yes/No more friendly in English though...

> When I pasted your last example it worked OK, I'm now getting HELLO printed. I'm almost there. Now I only need to find how to access current Fern root path (for example print current root path instead of HELLO).

Then you probably need to use `<Plug>(fern-action-tcd:root)` or whatever to change directory.

**Reply by lambdalisue** [reply]
> ls doesn't print plugin buffers. I only started looking into unlisted buffers because I noticed that after I spend some time navigating file system in Fern via enter (<CR>) & leave (<BS>) the amount of RAM used by nvim is starting to grow up a lot.
>
> What I discovered is that every time I enter into a folder or leave a folder this appears to be creating a new fern buffer instance and makes it active. Previous Fern buffers (all previous root folders I visited) remain open and I can switch to them via :b. This sounds more like a memory leak to me.
>
> There are more than 40 unlisted fern://drawer:1/file:///... buffers in my current session.

That's interesting while fern buffer should be unloaded automatically because `bufhidden` option is set to `unload`. Additionally, I know some people using fern daily but I haven't heralded about memory leak. 

What is your environment detail? OS, Vim/Neovim versions, CPU, Memory, or so on.

**Reply by iamvoidcoder** [reply]
This is Neovim 0.4.4 on MacOS 11.2.3

Latest Fern installed via vim-plug:

call plug#begin('~/.config/nvim/autoload/plugged')
    Plug 'lambdalisue/fern.vim' 
call plug#end()

Here is a simple test:

1. Starting nvim, new empty editing buffer:

```
:ls!
  2 %a   "[No Name]"                    line 1
```  

2. Opening an instance of Fern drawer:

```
:Fern . -drawer
:ls!
  2  a   "[No Name]"                    line 0
  3u%a-  "fern://drawer:1/file:///Users/Vcd;$" line 1
``` 

3. In the Fern drawer I'm entering directory "Temp" by pressing `<CR>`

```
:ls!
  2  a   "[No Name]"                    line 0
  3u# -  "fern://drawer:1/file:///Users/Vcd;$" line 13
  4u%a-  "fern://drawer:1/file:///Users/Vcd/Temp$" line 1
```  

As you can see the previous fern buffer remained opened. Now if I do :b3 it will correctly show the previous buffer. I can jump between these fern buffers and the buffers themselves are fully functional. However every time I do "enter" or "leave" command it will leave on of these unlisted buffers behind.

**Reply by lambdalisue** [reply]
Well, that's expected behavior. It should be OK.

> the amount of RAM used by nvim is starting to grow up a lot.

 I'd love to ask the memory usage. How much

**Reply by iamvoidcoder** [reply]
Ah this is expected then? How strange. So these hidden unlisted buffers, when are they going be freed? Is there a memory usage threshold or some automated garbage collector in vim to automatically free them? Sorry, I know nothing about the internals.

I haven't attempted any precise measurements. After starting nvim, with a single empty editing buffer open it is taking like 17MB. After opening Fern drawer and moving back and forth through the file system without opening any files I can see it jumping to about 40MB or so. 

This is not causing any issues really. I just wanted to understand why every leave/enter command is leaving an unneeded hidden buffer behind. Are these going to be freed eventually?

Thanks

**Reply by lambdalisue** [reply]
All I know is https://github.com/vim/vim/blob/dad4473f02e1fec86d43a2fc094536a4b27d3b25/runtime/doc/options.txt#L1343

If you care, use `setlocal bufhidden=wipe` would solve the issue but then you cannot get back with `<C-o>` or whatever.

**Reply by iamvoidcoder** [reply]
Thanks. This is sorting it. I think I'll go with an autocmd. 

autocmd FileType fern setl bufhidden=wipe

**Comment by oredaze** [comment]
I have a related question and it doesn't seem worth it to open a new topic.
I want to change the cwd once, only if I start vim with a directory (and thus start with fern).
Lets say I do `nvim ~/myproject/mydir` after that I always have to do `:cd ~/myproject/mydir`. I want to skip the second, but don't want any automatic cd-ing after that.

**Reply by lambdalisue** [reply]
Fern itself doesn't provide exact feature so you need to write some Vim script. See `:help VimEnter` for detail.
